File: problems/Palindrome Partitioning II/minCut.cs
public class Solution {
    public int MinCut(string s) {
        bool[,] isPalindromic = getIsPalindromicDp(s);

        return getMinCuts(s, isPalindromic);
    }

    private int getMinCuts(string s, bool[,] isPalindromicDp) {
        int n = s.Length;
        int[] dp = new int[n];

        for (int start = n - 1; 0 <= start; --start) {
            int minCuts = n;
            for (int end = n - 1; start <= end; --end) {
                if (isPalindromicDp[start, end]) {
                    minCuts = (n - 1 == end ? 0 : Math.Min(minCuts, 1 + dp[1 + end]));
                }
            }
            dp[start] = minCuts;
        }

        return dp[0];
    }

    private bool[,] getIsPalindromicDp(string s) {
        int n = s.Length;
        bool[,] dp = new bool[n, n];

        for (int start = n - 1; 0 <= start; --start) {
            for (int end = start; n > end; ++end) {
                if (s[start] == s[end] && (2 >= 1 + end - start || dp[1 + start, end - 1])) {
                    dp[start, end] = true;
                }
            }
        }

        return dp;
    }
}

File: problems/Palindrome Partitioning II/minCut.java
class Solution {
    public int minCut(String s) {
        int n = s.length();
        Boolean[][] isPalindromeDp = new Boolean[n][n];
        Integer[][] minCutsDp = new Integer[n][n];

        return getMinCutsTopDown(s, minCutsDp, isPalindromeDp, 0, s.length() - 1);
    }

    private int getMinCutsTopDown(String s, Integer[][] minCutsDp, Boolean[][] isPalindromeDp, int start, int end) {
        if (start >= end || isPalindrome(s, isPalindromeDp, start, end)) {
            return 0;
        }

        if (null == minCutsDp[start][end]) {
            int minCuts = end - start;

            for (int idx = start; end >= idx; ++idx) {
                if (isPalindrome(s, isPalindromeDp, start, idx)) {
                    minCuts = Math.min(minCuts, 1 + getMinCutsTopDown(s, minCutsDp, isPalindromeDp, 1 + idx, end));
                }
            }

            minCutsDp[start][end] = minCuts;
        }

        return minCutsDp[start][end];
    }

    private boolean isPalindrome(String s, Boolean[][] dp, int start, int end) {
        if (null == dp[start][end]) {
            int i = start;
            int j = end;

            dp[start][end] = true;

            while (i < j) {
                if (s.charAt(i++) != s.charAt(j--)) {
                    dp[start][end] = false;
                    break;
                }

                if (i < j && null != dp[i][j]) {
                    dp[start][end] = dp[i][j];
                    break;
                }
            }
        }

        return dp[start][end];
    }
}